### PR TITLE
Harmony 571 daac regression tests

### PR DESCRIPTION
**Pull request by asteiker**
Updated the regression notebook in the Harmony repo (HarmonyRegression.ipynb) and copied  into /test/harmony-regression. Added necessary `harmony_host_url` variable and modified each test to ensure compatibility across providers. Also added a few lines to test intake-stac compatibility of zarr outputs but left direct access of outputs commented out due to known issues with stac (https://github.com/intake/intake-stac/issues/48). A few assumptions:
1. I cannot test in SIT but I believe SIT points to UAT data so the collections that are tested are the same across SIT and UAT. 
2. I was able to build and test the environment.yaml locally but did not do a final check using the local docker instructions in the Readme (though I had tested this successfully with an intermediate notebook draft. 
3. I am not able to test in aws due to ngap issues with my account when ssm'ing but James had successfully tested with an intermediate notebook draft. 

This branch also has several updates and fixes that James completed including updates to the notebook helper functions under /test/harmony-regression to get some of the plotting working in the docker environment, and fixed a timeout issue to enable aws testing.

### CODE REVIEW

**Review by indiejames — APPROVED** [review]
Tested locally, in SIT, and UAT.

File: dot_env [hunk 1]
```diff
@@ -0,0 +1,6 @@
+REGRESSION_TEST_OUTPUT_BUCKET=<some bucket like harmony-sit-regression-tests>
```
[the remark this answers is not in the document]
Comment (bilts) [reply dot_env:1]: Also, since it's not used except in the deployed environment (i.e. not sourced by run_notebooks.sh) I think moving it to the script/ folder may solve the qualms here

File: script/deploy-from-docker.sh [hunk 2]
```diff
@@ -30,11 +30,15 @@ function retry {
   return 0
```
Comment (bilts) [inline script/deploy-from-docker.sh:30]: Not part of this PR, but for public consumption, this folder may be better off called something different and possibly containing its own README

File: script/test.sh [hunk 3]
```diff
@@ -12,6 +12,36 @@ function get_elb {
   echo $(aws elbv2 describe-load-balancers | jq --arg host "harmony-$HARMONY_ENVIRONMENT-frontend" '.LoadBalancers[] | select(.LoadBalancerName == $host) | .DNSName' | tr -d '"')
 }
 
+cd ..
```
Comment (bilts) [inline script/test.sh:15]: Again not part of this PR, but it's making me think we should consider pulling harmony-ci-cd to the open source side of the world and sticking these scripts in a more common location, given there are big pieces copy/pasted.

File: sshconfig [hunk 4]
```diff
@@ -4,3 +4,4 @@ Host i-*
     StrictHostKeyChecking no
     UserKnownHostsFile /dev/null
     LogLevel ERROR
+    ServerAliveInterval 60
```
Comment (bilts) [inline sshconfig:7]: Running locally, I needed a "PatrickAliveInterval" after about 7 minutes of no output :-/

File: test/harmony-regression/notebook_helpers/__init__.py [hunk 5]
```diff
@@ -0,0 +1,377 @@
+from contextlib import contextmanager
```
Comment (bilts) [inline test/harmony-regression/notebook_helpers/__init__.py:1]: There are possibly parts of this we could trim down or exclude, based on not caring much about the visual output in the current notebooks.  I'd vote for leaving it as is for now but revisiting once we revamp the existing notebooks to use the newly-developed Python lib.

Comment (bilts) [reply test/harmony-regression/notebook_helpers/__init__.py:1]: Doing the trimming down could remove a number of hefty dependencies in the current code

File: test/harmony-regression/notebook_helpers/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,11 @@
+requests >= 2.23.0
```
Comment (bilts) [inline test/harmony-regression/notebook_helpers/requirements.txt:1]: Is this completely replaced by conda now?  Can it be deleted?

File: test/run_notebooks.sh [hunk 7]
```diff
@@ -1,11 +1,39 @@
...
+  EXIT_CODES+=(${code})
+  NAMES+=(${name})
+done
+
+# check to see if any of the docker runs errored
+i=0
+exit_code=0
+for code in ${EXIT_CODES[@]}; do
+  name=${NAMES[$i]}
+  if [[ ${code} -ne 0 ]]; then
```
Comment (bilts) [inline test/run_notebooks.sh:32]: Add logs to note that the notebook is green, too, and possibly indicate that it has started.  Right now, on success, we get "Running regression tests", no output, a multi-minute-pause, then a zero exit code, which is very confusing and left me thinking something broke.

**Review by bilts — CHANGES_REQUESTED** [review]

File: sshconfig (hunk 4 above)
Comment (indiejames) [reply sshconfig:7]: papermill does not send output to stdout. remember that this is intended to be run in CI.

File: test/Dockerfile [hunk 8]
```diff
@@ -1,16 +1,21 @@
...
+RUN pip install conda-lock
+RUN conda install conda-lock
 
-COPY notebooks ./notebooks
-COPY harmony ./harmony
-COPY run_notebooks.sh .
+COPY .netrc .netrc
+RUN mkdir ./${sub_dir}
+COPY ${sub_dir}/conda-linux-64.lock ./${sub_dir}
+RUN ls ${sub_dir}
```
[the remark this answers is not in the document]
Comment (indiejames) [reply test/Dockerfile:17]: done

File: README.md [hunk 9]
```diff
@@ -2,58 +2,112 @@
...
-
-By default this will run the tests against the UAT environment. To run
-against a specific environment:
-
-    $ make run environment=prod
-
-Valid environment values are: sbx, sit, uat, prod.
-
-# Notebook Development
+    $ make images
```
[the remark this answers is not in the document]
Comment (indiejames) [reply README.md:16]: done

File: README.md [hunk 10]
```diff
@@ -2,58 +2,112 @@
...
+    $ cd script
+    $ export HARMONY_ENVIRONMENT=<uat|sit|sandbox|prod>
+    $ ./test.sh
 
-    $ ./run_notebooks harmony_host_url=<url of Harmony in the target environment>
-    
-e.g., 
-    
-    $ ./run_notebooks harmony_host_url="https://harmony.sit.earthdata.nasa.gov"
+Output will be in the bucket specified with the `REGRESSION_TEST_OUTPUT_BUCKET` environment
```
[the remark this answers is not in the document]
Comment (indiejames) [reply README.md:50]: 1. From the README > First create a `.env` file in the top level directory by copying in the `dot_env` file and filling
in the proper values.
The dot_env file example has this var.
2. The output does not go to the local output directory when run in AWS (what this particular section is describing). You may have had output from a local run still in there.

[the remark this answers is not in the document]
Comment (bilts) [reply README.md:50]: Some wires got crossed; I think I was looking at an outdated version of the README somehow.  Re-reviewing, because there are other things in here that I hadn't seen.

File: test/harmony-regression/notebook_helpers/requirements.txt (hunk 6 above)
Comment (indiejames) [reply test/harmony-regression/notebook_helpers/requirements.txt:1]: yes

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (indiejames) [reply README.md:53]: Actually I might be wrong about the output...

File: README.md [hunk 11]
```diff
@@ -2,58 +2,112 @@
...
+They _should_ utilize the `harmony_host_url` global variable to communicate with Harmony
+or to determine the Harmony environment. This variable is set by `papermill` - see the 
+`Harmony.ipynb` for how to make use of this variable. More information can be found
+in the [papermill](https://papermill.readthedocs.io/en/latest/usage-parameterize.html)
+documentation on setting parameters.
+
+New test suites must be added to the `Makefile`. A new `name-image` target (where name is the name of
+the test suite) should be added (see the `harmony-image` example), and the new image target
+should be added as a dependency of the `images` target. The docker image should have a name like
+`harmony/regression-tests-<base_name>`, where `base_name` is the name of the test suite.
```
[the remark this answers is not in the document]
Comment (bilts) [reply README.md:109]: Filed a ticket

File: README.md [hunk 12]
```diff
@@ -2,58 +2,112 @@
...
+New test suites must be added to the `Makefile`. A new `name-image` target (where name is the name of
+the test suite) should be added (see the `harmony-image` example), and the new image target
+should be added as a dependency of the `images` target. The docker image should have a name like
+`harmony/regression-tests-<base_name>`, where `base_name` is the name of the test suite. 
+
+Finally, add the image base name to the `images` array on line 6 of the `run_notebooks.sh` file.
+For instance, if the image is named `harmony/regression-tests-foo`, then we would add `foo` to the
+array.
+
+The `run_notebooks.sh` file can be used as described above to run the test suite. Notebooks are
```
[the remark this answers is not in the document]
Comment (bilts) [reply README.md:115]: Ticket filed.

File: script/deploy-from-docker.sh (hunk 2 above)
Comment (bilts) [reply script/deploy-from-docker.sh:30]: (Nevermind)

File: script/test.sh (hunk 3 above)
Comment (bilts) [reply script/test.sh:15]: HARMONY-761 now

File: sshconfig (hunk 4 above)
Comment (bilts) [reply sshconfig:7]: It gets run in CI and by people who are testing notebooks to provide to us.  I'm thinking more about the latter.  I'm planning to add a couple of echo statements that will at least show something happening and indicate success.

File: test/harmony-regression/notebook_helpers/__init__.py (hunk 5 above)
Comment (bilts) [reply test/harmony-regression/notebook_helpers/__init__.py:1]: HARMONY-762

**Review by bilts — APPROVED** [review]
The latest updates address everything I had.
